cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -192,6 +195,60 @@ struct TemporalComponentExtract
   }
 };
 
+Status CheckTimezones(const ExecBatch& batch) {
+  const auto& timezone = GetInputTimezone(batch.values[0]);
+  for (int i = 1; i < batch.num_values(); i++) {
+    const auto& other_timezone = GetInputTimezone(batch.values[i]);
+    if (other_timezone != timezone) {
+      return Status::TypeError("Got differing time zone '", other_timezone,
+                               "' for argument ", i + 1, "; expected '", timezone, "'");
+    }
```

Comment:
What would it mean to find the difference between two timestamps in different timezones? I suppose we could compute the difference in UTC, but then you'd get different results (as compared to working with local timestamps) if you had a DST transition in the covered time.

Reply:
Difference between two moments is just an interval without a timezone. And out timezone aware timestamps are unambiguous moments or am I getting something terribly wrong here? :)

Reply:
Oh. Doh. Yes, you are right. Thanks for pointing this out :)

Reply:
Er, wait. So concretely, if I have these timestamps:

```
2019-12-31 18:00:00-0500
2019-12-31 23:00:00-0500
```

In America/New_York, the year difference is 0, but in UTC, the year difference is 1 (the latter timestamp is 04:00 on 2020-01-01). This kernel will localize the timestamps, so the expected difference is 0. But this kind of calculation doesn't make sense if the time zones are different, unless you suddenly switch to interpreting zoned timestamps as their underlying UTC value. I'm worried this is potentially confusing/a footgun.

Reply:
Riiight it's because you care about boundaries not the full periods :). In that case you'd want to block this indeed and have user change one of the two timestamps into the timezone they care about and run the operation 'there'.

Reply:
Sorry for the misdirect.

Reply:
Sorry for the misdirect.

Reply:
Ok, sounds good, sorry for the confusion. I realized I forgot to update the docs, I'll try to add some explanatory text there as well since this is a little confusing.